DESIGN.md

```diff
@@ -0,0 +1,79 @@
+
+For Authorization, a simple form of role based access control will be implemented: Each role will have its own CA. All CAs will be trusted by the server. Upon handling a remote procedure call, the server will first check the client certificate's certificate authority and compare it to its internal mapping of certificate authority to privilege level. The following roles will be implemented:
+- **Read only:** this role can get job output, but not create or delete jobs. This is useful for automated tooling such as status indicators and job notifiers.
+- **Read/Write access:** this role can get job output as well as create and delete jobs. This role is useful for administrators and any kind of deployment automation that would leverage this service.
+
+![visual representation of the multiple client CA model](doc/client_cas.png)
+
+## API
+### Endpoints
+The API exposed by the RPC service will contain the following endpoints
+- **CreateJob**: receives a Job and returns a CreateJobOutput
+- **DeleteJob**: receives a PID and returns a DeleteJobOutput
+- **ListJobs**: returns a JobList
+- **GetOutput**: receives a PID and returns a JobOutput
+
+### Datatypes
+- **Job**
+    - Command (string):
+    Command holds a valid shell command that is run on the underlying server shell.
+    - PID (integer):
+    PID holds an integer representing job ID. In this project it will correlate directly to the pid of the process running on the RPC server for a given job. In a production RPC system it is advisable to use a randomized PID.
+    - CreatedTime (string):
+    Timestamp that represents when the process was launched on the server
+- **JobList**
+    - Jobs ([]Job):
+    A list of Job obejcts as defined above in this list
+- **JobOutput**
+    - Command (string):
+    Holds the shell command as originally defined in the corresponding call to CreateJob
+    - CurrentOutput (string):
+    Current cummulative contents of stdout from Job
+    - Complete (boolean):
+    Boolean that represents whether the job is still running or if it has finished
+- **DeleteJobOutput**
+    - Job (Job):
+    Job object as defined above in this list
+    - Exit (integer):
+    Exit code as returned from the underlying process
+    - Interrupted (boolean):
+    Boolean is set to true if underlying process was still running when DeleteJob was called
+- **CreateJobOutput**
```

Comment:
We only need the id here and it would be better to leverage the built in error returned by the grpc call instead of baking it in the response

Reply:
I agree that we only need the ID here, but if we return the full job struct we can get early feedback back to the client if there is an immediate error or immediate results to return.

I have updated this section in the latest revision.